File: examples/python/gi/add_connection.py
#!/usr/bin/env python
# SPDX-License-Identifier: GPL-2.0+
#
# Copyright (C) 2014 Red Hat, Inc.
#

#
# This example shows how to add a new NM connection profile.
# The code uses libnm (NM) via GObject Introspection.
#
# Documentation links:
# https://developer.gnome.org/libnm/1.0/
# https://developer.gnome.org/NetworkManager/1.0/ref-settings.html
#

import gi
gi.require_version('NM', '1.0')
from gi.repository import GLib, NM
import sys, uuid

main_loop = None

def print_values(setting, key, value, flags, data):
    print("  %s.%s: %s" % (setting.get_name(), key, value))

# create an Ethernet connection and return it
def create_profile(name):
    profile = NM.SimpleConnection.new()
    s_con = NM.SettingConnection.new()
    s_con.set_property(NM.SETTING_CONNECTION_ID, name)
    s_con.set_property(NM.SETTING_CONNECTION_UUID, str(uuid.uuid4()))
    s_con.set_property(NM.SETTING_CONNECTION_TYPE, "802-3-ethernet")

    s_wired = NM.SettingWired.new()

    s_ip4 = NM.SettingIP4Config.new()
    s_ip4.set_property(NM.SETTING_IP_CONFIG_METHOD, "auto")

    s_ip6 = NM.SettingIP6Config.new()
    s_ip6.set_property(NM.SETTING_IP_CONFIG_METHOD, "auto")

    profile.add_setting(s_con)
    profile.add_setting(s_ip4)
    profile.add_setting(s_ip6)
    profile.add_setting(s_wired)

    print("Created connection profile:")
    profile.for_each_setting_value(print_values, None)

    return profile

# callback function
def added_cb(client, result, data):
    try:
        client.add_connection_finish(result)
        print("The connection profile has been successfully added to NetworkManager.")
    except Exception as e:
        sys.stderr.write("Error: %s\n" % e)
    main_loop.quit()

if __name__ == "__main__":
    # parse arguments
    persistent = False
    if len(sys.argv) != 2 and len(sys.argv) != 3:
        sys.exit('Usage: %s <connection name> [persistent]' % sys.argv[0])
    if len(sys.argv) == 3:
        if sys.argv[2] in "persistent" and sys.argv[2][:1] == "p":
            persistent = True
        else:
            sys.exit('Usage: %s <connection name> [persistent]' % sys.argv[0])
    profile_name = sys.argv[1]

    main_loop = GLib.MainLoop()

    # create Client object
    client = NM.Client.new(None)

    # create a connection profile for NM
    con = create_profile(profile_name)

    # send the connection to NM
    client.add_connection_async(con, persistent, None, added_cb, None)

    main_loop.run()

File: examples/lua/lgi/add-connection.lua
#!/usr/bin/env lua
-- SPDX-License-Identifier: GPL-2.0-or-later
--
-- Copyright (C) 2015 Red Hat, Inc.
--

-- Adding an Ethernet connection to NetworkManager in Lua.
-- The example uses libnm library using GObject introspection via Lua lgi module.
-- Most distribution ship the module as lua-lgi package.
-- libnm guide:   https://developer.gnome.org/libnm/1.0/
-- Lua-lgi guide: https://github.com/pavouk/lgi/blob/master/docs/guide.md

local lgi = require 'lgi'
local GLib = lgi.GLib
local NM = lgi.NM

function uuid()
  math.randomseed(os.time())
  local template ='xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'
  local uuid = string.gsub(template, '[xy]', function (c)
    local v = (c == 'x') and math.random(0, 0xf) or math.random(8, 0xb)
    return string.format('%x', v)
  end)
  return uuid
end

-- function creating NMConnection
function create_profile(name)
  profile = NM.SimpleConnection.new()

  s_con = NM.SettingConnection.new()
  s_wired = NM.SettingWired.new()
  s_con[NM.SETTING_CONNECTION_ID] = name
  s_con[NM.SETTING_CONNECTION_UUID] = uuid()
  s_con[NM.SETTING_CONNECTION_TYPE] = "802-3-ethernet"

  profile:add_setting(s_con)
  profile:add_setting(s_wired)

  -- show the connection
  -- profile:dump()
  return profile
end

-- callback function for add_connection_async()
function added_cb(client, result, data)
  local con,err,code = client:add_connection_finish(result)
  if con then
    print("The connection profile has been successfully added to NetworkManager:")
    print(con:get_id(), con:get_uuid())
  else
    print(string.format("Error: (%d) %s", code, err))
  end
  main_loop:quit() -- exit now
end


---------------------------
-- Main code starts here --
---------------------------
-- parse command-line arguments
local name, persist = ...
if (not name or (persist and not string.find("persistent", persist, 1))) then
  print(string.format("Usage: %s <connection name> [persistent]", arg[0]:gsub(".*/","")))
  os.exit(1)
end

-- create GLib main loop
main_loop = GLib.MainLoop(nil, false)

-- create Client object
local client = NM.Client.new()

-- create a connection profile
local con = create_profile(name)

-- send the connection to NetworkManager
client:add_connection_async(con, persist, nil, added_cb, nil)

-- run main loop so that the callback could be called
main_loop:run()

